### DISCUSSION TITLE
Possible error handling missing or docs update required

### DISCUSSION THREAD

**Discussion by eugeneniemand** [tsightler/ring-mqtt#852] [General Discussion]
I reinstalled my entire RaspberryPi OS on Sunday which meant latest Docker etc. When running ring-mqtt it would authenticate but hang at device discovery. I dont have anything online. After investigating some other issues I had and digging through `journalctl` logs I noticed this in the logs 1000s of time

`Jun 12 08:35:13 raspberrypi dockerd[774]: time="2024-06-12T08:35:13.682751142+01:00" level=error msg="[resolver] failed to query external DNS server" client-addr="udp:172.18.0.3:60328" dns-server="udp:192.168.1.3:53" error="read udp 172.18.0.3:60328->192.168.1.3:53: i/o timeout" question=";oauth.ring.com.\tIN\t A" spanID=670b06614735068a traceID=98a1768b9c59a0b8678644fa8a529c45`

I then found another GitHub issue referencing the below deprecation in Docker v26

`CVE-2024-29018
: Do not forward requests to external DNS servers for a container that is only connected to an 'internal' network. Previously, requests were forwarded if the host's DNS server was running on a loopback address, like systemd's 127.0.0.53. moby/moby#47589`

I had to add `dns: 1.1.1.1` to my docker compose and this resolved the Device discovery hanging issue and everything started working again. By reading the above deprecation I could possibly also use `network_mode: "host"` but havent tried. It is odd though that the util to get the token works and that the login in the container start-up works but then hangs on device discovery. There are no errors in the container logs

**Comment by tsightler** [comment]
I'm pretty sure this is just a host configuration issue.  You posted references to CVE-2024-29018, but that is irrelevant because it is specific to using internal networks, such as using `internal: true` in Docker Compose, but this creates a truly internal only network, i.e. a network without any external connectivity at all, and ring-mqtt can't run in such a configuration in any case.  That's actually what the CVE is about, a container running on an internal only network was leaking data externally via forwarded DNS queries when there should have been no external network at all.

IMO, the original error message is very clear, it shows that dockerd is attempting to forward the DNS request from the container to a local DNS server at 192.168.1.3 and this query is timing out.  Why it is timing out?  Well, that I can't answer, but adding `dns: 1.1.1.1` is simply a way to tell dockerd to bypass the local hosts resolver and use a different forwarder for that container, so sure, that works, but it will also dramatically increase your outbound DNS resolution traffic as ring-client-api makes tons of DNS queries and you are far better off using a local, caching DNS.  I would suggest investigating why your local DNS resolver is not responding to the forwarded query.
